### PR TITLE
[Hims] Add SPP connection [BrailleEdge]

### PR DESCRIPTION
**Pull request by EdKweon**
### Summary of the issue:
Support for spp connection with a Braille Edge device.

### Description of user facing changes
Support for all BrailleEdge devices NVDA's usb connection.

### Description of development approach
Support for Braille Edge device SPP connection in the Hims display. Only BrailleEdge devices with changed hardware chips can connect to SPP(Not use Hims Driver). Original BrailleEdge should use Hims driver to connect.

### Testing strategy:
How have you tested to ensure that your change works as intended?
yes.
Have you ensured testing coverage across all supported operating systems?
yes.
Have you considered possible regressions (related features or behaviours that may break)?
yes.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

### CODE REVIEW

File: source/brailleDisplayDrivers/hims.py [hunk 1]
```diff
@@ -190,12 +190,13 @@ def registerAutomaticDetection(cls, driverRegistrar: bdDetect.DriverRegistrar):
 		# Bulk devices
 		driverRegistrar.addUsbDevices(bdDetect.DeviceType.CUSTOM, {
 			"VID_045E&PID_930A",  # Braille Sense & Smart Beetle
-			"VID_045E&PID_930B",  # Braille EDGE 40
+   			"VID_045E&PID_930B",  # Braille EDGE 40
```
Comment (LeonarddeR) [inline source/brailleDisplayDrivers/hims.py:193]: Please undo the addition of spaces

```suggestion
			"VID_045E&PID_930B",  # Braille EDGE 40
```

File: source/brailleDisplayDrivers/hims.py [hunk 2]
```diff
@@ -190,12 +190,13 @@ def registerAutomaticDetection(cls, driverRegistrar: bdDetect.DriverRegistrar):
...
 		driverRegistrar.addUsbDevices(bdDetect.DeviceType.CUSTOM, {
 			"VID_045E&PID_930A",  # Braille Sense & Smart Beetle
-			"VID_045E&PID_930B",  # Braille EDGE 40
+   			"VID_045E&PID_930B",  # Braille EDGE 40
 		})
 
 		# Sync Braille, serial device
 		driverRegistrar.addUsbDevices(bdDetect.DeviceType.SERIAL, {
 			"VID_0403&PID_6001",
+   			"VID_1A86&PID_55D3",  # Braille Emotion 40
```
Comment (LeonarddeR) [inline source/brailleDisplayDrivers/hims.py:199]: Please remove the spaces here as well
```suggestion
			"VID_1A86&PID_55D3",  # Braille Emotion 40
```

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
Could you please elaborate on how you tested this? I think this pr is not complete yet. You might need to add an extra model class and add it to the modelMap.

**Comment by EdKweon** [comment]
> Could you please elaborate on how you tested this? I think this pr is not complete yet. You might need to add an extra model class and add it to the modelMap.

First of all I'm sorry for not checking the code more accurately.
I tested that new device can get NVDA braille data in connectivity app. And check unit test using `rununittests.bat`. 

I agree with you that new model class and that should be added in the modelMap. I'll add your feedback code.  Any other things I should check?

**Comment by LeonarddeR** [comment]
Apart from the model map, I guess this is ok now.

**Comment by EdKweon** [comment]
I added device to model map. Thank you for your review.

**Comment by EdKweon** [comment]
I added blank line because of Lint error. I don't know why this happend. Please review my code please

**Comment by seanbudd** [comment]
Hi @EdKweon,

It appears this code is just for internal testing for your organisation, rather than for wider NVDA users.
As such, I don't think we can merge this until it is a complete feature targeted at wider NVDA users.

If you need signed builds to test with, we are happy to create them - I've created a signed build from this PR - [download](https://ci.appveyor.com/api/buildjobs/3p8c339mkhgph3fc/artifacts/output/nvda_snapshot_try-test-HIMS-30691,ee84dcd8.exe).

You can also create your own [self signed builds](https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/selfSignedBuild.md) for testing if you don't want to wait on us.

Feel free to push further changes to this PR and ask us to create another signed build if you need to make any further changes for testing.
When this PR is ready for the public, feel free to mark it as ready for review to convert it from a draft.

Thanks,
Sean

**Comment by michaelDCurran** [comment]
@EdKweon  I think there is some confusion here. Could you please clarify if the BrailleEmotion is a real-world Braille display on the market, or it is just an emulated test device? If it is a real Braille display on the market, of course we will accept this PR. 
If it is a device / emulator purely for testing, then we cannot take this code.
Apologies if we have miss-understood the intention of this PR.

**Comment by EdKweon** [comment]
@michaelDCurran 
This is a device that will actually be sold on the market. Of course, the code for that device may be PR with additional requirements and testing. It is not simply a device for testing.

**Comment by seanbudd** [comment]
Thanks for more information, please let us know when testing is complete and the pull request is complete

**Comment by EdKweon** [comment]
@seanbudd
 I think first PR and  test is complete.

File: source/brailleDisplayDrivers/hims.py [hunk 3]
```diff
@@ -116,6 +116,11 @@ def _get_keys(self):
 		})
 		return keys
 
+
+class BrailleEmotion(BrailleEdge):
+	name = "Braille Emotion"
+	usbId = "VID_1A86&PID_55D3"
+
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/hims.py:123]:
```suggestion


```

**Review by seanbudd — COMMENTED** [review]
Please add a reference to the device in the english user guide with the other HIMS devices, so our users know that the device works with NVDA.

Please also add a changelog entry in changes.t2t to confirm that the new device will work in the 2024.2 release of NVDA.

Has the device been fully tested? Are there any expected further changes.
Many of the other drivers define more keys, will that be necessary here?

**Comment by seanbudd** [comment]
Could you please also update the PR description to be more accurate.
The PR description implies this is for internal testing for your company.
Rather it should be centred towards our users, i.e. "Add support for Braille eMotion"

**Comment by EdKweon** [comment]
@seanbudd 
I updated PR description. Thanks for advice!

**Comment by seanbudd** [comment]
Hi - this still needs an update to the change log and user guide

**Comment by josephsl** [comment]
Note: English translation to follow:

안녕하세요,

NVDA 개발 공헌자 이성원입니다. 요즈음 건강하시지요?

먼저 새 힘스 기기를 NVDA 스크린 리더 사용자들이 사용할 수 있도록 본 PR을 보내주셔서 감사합니다. 선생님과 NV Access 맴버들간 주고받는 내용을 보고 아마도 한국인이신것 같고 또 이 프로잭트에 처음 PR을 보내주시는것 같아 이 포스트를 쓰게 되었습니다.

제가 이해하기에:

1. 추후 나올 힘스 제품 관련 PR: yes
2. 곧 상용화 또는 한국 스크린 리더(예: 센스리더) 외 타 스크린 리더 지원 준비를 위한 PR: yes

이렇게 이해하면 됩니까?

또한 점자기기 지원 PR은 이미 출시된 제품 또는 상용화될 제품에 한하여 PR을 받고있는데 보통 기기 사용자나 제품 회사에서 작성하게 됩니다. 예전에는 지원 코드 및 사용설명서(user guide)에 제품 관련 정보(모델명, 통신 방법, 명령 세트 등)은 PR 작성자가 적고 변경 내역(changelog)은 NV Access가 작성했는데 작년 말부터 후자도 PR 작성자가 작성하도록 변경되었습니다. 만약 사용설명서나 변경 내역에 대해 번역이 필요하시면 연락해 주세요.

새 힘스 기기 지원을 위한 PR을 보내주셔서 다시 한번 감사드립니다.

English translation:

Hello,

I am Joseph Lee, one of the volunteer code contributors for NVDA screen reader. hope you are staying sare and healthy these days.

Thank you for offering a pull request to add support for a new HIMS device. Based on the conversation I read, I felt you are a Korean and a first-time contributor, so I decided to post a short explanation.

If I understood the current PR state right:

1. This is a PR to support an upcoming HIMS braille device: yes
2. This PR expands support for the upcoming or a just released HIMS device beyond the Korean screen reader market (e.g. Sense Reader): yes

Am I correct?

Also, typically braille device support is provided by the user or the manufacturer for a device already in production or shortly to be released. Until late last year, PR's for braille device support was done in two parts: code and user guide section written by the PR writer, and NV Access folks writing what's new document. Today, all three parts - code, user uide, what's new entry - are written by PR writers. Please let me know if you need translations for user guide and/or the changelog from Korean to English.

Again thnk you for the PR to add an upcoming HIMS device for use by NVDA users.

End of translation

Hope this makes sense. Thanks.

**Comment by EdKweon** [comment]
@josephsl 
우선 아마 사용설명서에  모델명, 통신방법, 명령어 세트 까지 작성해서 PR이 요청되야 한다면, device에 관한 코드를 좀 더 명확히 작성 후 review 신청하도록 하겠습니다.

- 추후 나올 힘스 제품 관련 PR: yes
-곧 상용화 또는 한국 스크린 리더(예: 센스리더) 외 타 스크린 리더 지원 준비를 위한 PR: yes

두가지 사항 모두 맞습니다. 추후 PR사용설명서(user guide), 변경 내역(changelog) 작성 후 오작성 되거나 오해의 소지가 있는 부분이 있는지 요청드려도 될까요? NVDA PR을 하는 것이 처음이라 과정들이 아직 익숙치가 않습니다.  


감사합니다.

English Translation:

First of all, if a PR is requested by writing the model name, communication method, and command set in the user manual, I will write the code for the device more clearly and then apply for review.

- PR related to Hims products to be released in the future: yes
-PR for preparation for upcoming commercialization or support for screen readers other than Korean screen readers (e.g. Sense Reader): yes

Both points are correct. After writing the PR user guide and change log in the future, may I ask if there are any incorrect or misleading parts? This is my first time doing NVDA PR, so I'm not familiar with the process yet.

Thanks.

**Comment by josephsl** [comment]
안녕하세요,

알겠습니다. 만약 시간이 있으시면 NVDA 한국 사용자들과 NVDA 개발 및 이 PR에 대해서 이야기를 나누어 보시면 좋겠다고 생각합니다(참고로 저는 미국에 있고 현재 대학원생이라 자세한 코드 리뷰는 해드릴 수 없지만). 도움이 되셨으면 합니다.

English: Got it. If you have time, I think it would be great to talk to Korean NVDA users about the screen reader devleopment process and this PR (I'm a graduate student in United States so can't review the PR in detail at the moment). Hope this helps.

Tanks.

**Comment by EdKweon** [comment]
This may be misleading from what I said before. I will summarize the current situation in the article below. Currently, the device I am trying to add is almost same as the BrailleEdge product. However, hardware was little bit changed. So the product is connected to nvda via serial (not use Hims Driver). Therefore, there is no key to add or modify separately in the user_docs document.

Thanks for members.

File: source/brailleDisplayDrivers/hims.py [hunk 4]
```diff
@@ -116,6 +116,24 @@ def _get_keys(self):
 		})
 		return keys
 
+
+class BrailleEdge2S(BrailleEdge):
+	"""This device is BrailleEdge which doesn't use hims driver. It only use spp connection.
+	"""
+	usbId = "VID_1A86&PID_55D3"
+
+	def _get_keys(self):
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/hims.py:125]: please add a return type here

```suggestion
	def _get_keys(self) -> dict[str, str]:
```

File: source/brailleDisplayDrivers/hims.py [hunk 5]
```diff
@@ -116,6 +116,24 @@ def _get_keys(self):
 		})
 		return keys
 
+
+class BrailleEdge2S(BrailleEdge):
+	"""This device is BrailleEdge which doesn't use hims driver. It only use spp connection.
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/hims.py:121]:
```suggestion
	"""This device is the BrailleEdge which doesn't use the hims driver.
	It only uses a SPP connection.
```

File: user_docs/en/changes.t2t [hunk 6]
```diff
@@ -7,9 +7,9 @@ What's New in NVDA
 
 == New Features ==
 - In Windows 11, NVDA will announce alerts from voice typing and suggested actions including the top suggestion when copying data such as phone numbers to the clipboard (Windows 11 2022 Update and later). (#16009, @josephsl)
+- NVDA will now automatically detect the BrailleEdgeS2 devices via USB. (#16033)
 -
 
-
```
Comment (seanbudd) [inline user_docs/en/changes.t2t:12]: please restore this removed new line

File: user_docs/en/changes.t2t [hunk 7]
```diff
@@ -7,9 +7,9 @@ What's New in NVDA
 
 == New Features ==
 - In Windows 11, NVDA will announce alerts from voice typing and suggested actions including the top suggestion when copying data such as phone numbers to the clipboard (Windows 11 2022 Update and later). (#16009, @josephsl)
+- NVDA will now automatically detect the BrailleEdgeS2 devices via USB. (#16033)
```
Comment (seanbudd) [inline user_docs/en/changes.t2t:10]:
```suggestion
- Added support for the BrailleEdgeS2 braille device. (#16033)
```

**Review by seanbudd — COMMENTED** [review]
Can you please also document this device in the [user guide](https://www.nvaccess.org/files/nvda/documentation/userGuide.html#Hims)?

**Comment by EdKweon** [comment]
@seanbudd 
There are no other keys to add or modify in user_guide. Because this  is completely same with BrailleEdge except not using Hims driver(hardware chip replacement).

**Comment by seanbudd** [comment]
This is a new braille device right? We like to include all of the supported braille devices in the user guide.

**Comment by EdKweon** [comment]
@seanbudd 
What I said before was incorrect information. It's just a BrailleEdge not a new device Braille eMotion.

**Review by seanbudd — APPROVED** [review]
Thanks for explaining

**Comment by seanbudd** [comment]
can you please update the PR description to be more accurate? thanks

**Comment by EdKweon** [comment]
@seanbudd 
Thanks. I updated PR description.

**Comment by seanbudd** [comment]
Great - when this is merged it will become available in NVDA alpha releases. As long as no major issues are found it will be in NVDA 2024.2

File: source/brailleDisplayDrivers/hims.py [hunk 8]
```diff
@@ -116,6 +116,25 @@ def _get_keys(self):
...
 		return keys
 
+
+class BrailleEdge2S(BrailleEdge):
+	"""This device is the BrailleEdge which doesn't use the hims driver.
+	It only uses a SPP connection.
+	"""
+	usbId = "VID_1A86&PID_55D3"
+
+	def _get_keys(self) -> dict[str, str]:
```
Comment (LeonarddeR) [inline source/brailleDisplayDrivers/hims.py:126]: If the keys on the new Edge are exactly the same as the keys on the old device, I think you can remove this override.

Comment (EdKweon) [reply source/brailleDisplayDrivers/hims.py:126]: I think my post lacked explanation. Same device means focus to users. In the device, key packet data little bit changed.

**Review by LeonarddeR — APPROVED** [review]

**Pull request merged**